### ISSUE TITLE
bcachefs format will allow me to add all the copies to one physical device

### ISSUE THREAD

**Issue by b-r-o-w-n** [koverstreet/bcachefs#144]
bcachefs format will allow unsafe configurations.

I have 4 partitions on /dev/sda: sda1, sda2, sda3 and sda4.
I can do this unsafe configurations:

bcachefs format  --group hdd /dev/sda1  /dev/sda2 /dev/sda3 /dev/sda4  --replicas=3 --data_replicas_required=3 --metadata_replicas_required=3
mount -t bcachefs  -o noatime /dev/sda1:/dev/sda2:/dev/sda3:/dev/sda4   /mnt/bcachefs1

I think it might be wise for the command to refuse this unsafe configuration.

**Comment by godofdream** [comment]
Other format clis don't check if it's the same disk (e.g. btrfs).

Also there is actually a usecase (not a good one) for allowing multiple partitions on one disk.
you could partition around bad sectors and use the partitions as `bcachefs format --group hdd /dev/sda1 /dev/sda2 /dev/sda3 /dev/sda4 --replicas=1`

same occurs if you use some kind of ramdisk, lvm etc.

**Comment by b-r-o-w-n** [comment]
Yes, for testing adding 3 replicas on a single drive saves me having to find two additional drives.
I'll get back to you on this... I want to see that btrfs does;) zfs would be interesting to check also.

**Comment by nckx** [comment]
> I think it might be wise for the command to refuse this unsafe configuration.

It might merit a warning but please don't make it an error (or hide it behind one big `--force` flag that could mask real issues).

I use it in several tests.  There are other perfectly valid use cases for this if you know something that `bcachefs` doesn't.

**Comment by YellowOnion** [comment]
you can set a device's durability to 0 during format/add.

A warning or (bypassable) error might be smart.

**Comment by Komzpa** [comment]
Opened a tools PR for the small guardrail version of this:

https://github.com/koverstreet/bcachefs-tools/pull/666

Companion ktest proof:
https://github.com/koverstreet/ktest/pull/87

It warns during `bcachefs format` when the requested data/metadata replica count is greater than the number of physical parent disks found via sysfs, and lists the format devices that share a parent disk. I left it as a warning instead of a hard error because this thread also has legitimate partition/test use cases. `--quiet` suppresses the warning.

Validated with:
- `git diff --check`
- `BINDGEN=/home/kom/.cargo/bin/bindgen make -j32 bcachefs`
- `bash -n tests/fs/bcachefs/format.ktest` in the companion ktest PR

**Comment by koverstreet** [comment]
Landed in the testing branch as 2c5c20a2eb6d ("format: warn when replicas share a disk"), merged from PR koverstreet/bcachefs-tools#666 — we merge via cherry-pick, so GitHub's auto-close didn't fire. Closing manually.
